File: Data Structures & Algorithms/interleaving-string/submission-3.py
from collections import defaultdict

class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False

        memo = defaultdict(bool)

        def helper(p1, p2):
            p3 = p1 + p2

            if p3 == len(s3):
                return p1 == len(s1) and p2 == len(s2)

            if (p1, p2) in memo:
                return memo[(p1, p2)]

            res = False
            if p1 < len(s1) and s1[p1] == s3[p3]:
                res |= helper(p1 + 1, p2)
            if p2 < len(s2) and s2[p2] == s3[p3]:
                res |= helper(p1, p2 + 1)
            memo[(p1, p2)] = res
            return memo[(p1, p2)]

        return helper(0, 0)
            

File: Data Structures & Algorithms/climbing-stairs/submission-3.java
class Solution {
    public int helper(int i, int n, Map<Integer, Integer> memo){
        if(i > n)
            return 0;
        if(i == n)
            return 1;
        if(memo.containsKey(i))
            return memo.get(i);
        memo.put(i, memo.getOrDefault(i, helper(i + 1, n, memo) + helper(i + 2, n, memo)));
        return memo.get(i);
    }

    public int climbStairs(int n) {
        Map<Integer, Integer> memo = new HashMap<>();
        return helper(0, n, memo);
    }
}
